New optional parameters for course blocks API: lti_url, block_types

lenacom opened the pull request:
This is a continuation of the pull request https://github.com/edx/edx-platform/pull/10951

**Background:** Business people need to generate a list of LTI URLs for a given course, to provide to a customer. This new request should reduce the time and effort needed to pull LTI endpoints for a course, and reduce the chance of human error. 

**Studio Updates:** None.

**LMS Updates:** 2 new optional parameters are added to /api/courses/v1/blocks/?course_id=<course_id>: 
- 'requested_fields' can accept 'lti_url' (indicates it's required to show the field 'lti_url');
- 'block_types' (a list of requested block types; possible values include sequential, vertical, html, problem, video, and discussion). 

**Manual test instructions:**
Open in the browser:
http://127.0.0.1:8000/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&username=staff&depth=all&block_counts=video&block_types=html,video&requested_fields=lti_url
![edx23](https://cloud.githubusercontent.com/assets/7777978/12007611/bb5b91d4-ac1c-11e5-8eb3-775f2dedd084.jpg)

edx-webhook commented:
Thanks for the pull request, @lenacom! I've created [OSPR-1168](https://openedx.atlassian.net/browse/OSPR-1168) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -20,6 +20,7 @@ def get_blocks(
         block_counts=None,
         student_view_data=None,
         return_type='dict',
+        block_types=None,
```
You should add `block_types` to the docstring.

lenacom replied:
Added block_types to the docstring below.

nedbat commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -61,6 +62,18 @@ def get_blocks(
     blocks = get_course_blocks(user, usage_key, transformers)
 
+    # filter blocks by types
+    if block_types and len(block_types) == 0:
```
I don't know what type `block_types` will be, but empty containers in Python are false, so this expression is likely never going to be true.

nedbat commented on lms/djangoapps/course_api/blocks/tests/test_serializers.py:
```diff
@@ -136,9 +138,6 @@ def create_serializer(self):
         serializer = self.create_serializer()
 
-        # verify root
-        self.assertEquals(serializer.data['root'], unicode(self.block_structure.root_block_usage_key))
```
Removing assertions from tests makes me nervous :)

Can you explain why we no longer want this check?

lenacom replied:
I replaced 'return self.topological_traversal()' with 'return self.get_block_keys()' in block_structure.py:BlockStructure
as @nasthagiri suggested. So now blocks are treated as a list and no root is assigned.

nedbat commented:
@lenacom there's also one small pep8 thing to fix: use `x not in y` instead of `not x in y`.

nedbat commented:
@robrap @nasthagiri a continuation...

lenacom commented:
jenkins run bokchoy

robrap commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -44,6 +45,8 @@ def get_blocks(
         return_type (string): Possible values are 'dict' or 'list'. Indicates
             the format for returning the blocks.
+        block_types (list): Optional list of names of block types which to
+            return blocks.
```
Can you update to something like this?
```
block_types (list): Optional list of block type names used to filter the final result of returned blocks.
```
@nasthagiri Is this a reasonable description?

nasthagiri replied:
Yes. I agree with @robrap's suggestion.

robrap commented on lms/djangoapps/course_api/blocks/serializers.py:
```diff
@@ -70,7 +77,6 @@ class BlockDictSerializer(serializers.Serializer):  # pylint: disable=abstract-m
     Serializer that formats a BlockStructure object to a dictionary, rather
     than a list, of blocks
     """
-    root = serializers.CharField(source='root_block_usage_key')
```
@nasthagiri Any idea why this was here?

robrap replied:
@lenacom I now saw your comment to @nedbat below.  I'll have @nasthagiri review to make sure this is as she expected.

nasthagiri replied:
We need to keep `root` in this serializer.  It indicates the `root` of the block structure.  When returning a `list` of blocks, using the `BlockSerializer`, the `root` and `blocks` fields do not exist.

robrap commented on lms/djangoapps/course_api/blocks/tests/test_api.py:
```diff
@@ -77,3 +75,17 @@ def test_sub_structure(self):
+                ('problem', 'problem_y1a_3'),
+        ):
+            block = self.store.get_item(self.course.id.make_usage_key(block_type, block_name))
+            self.assertIn(unicode(block.location), blocks['blocks'])
```
I think this test should show (through code or comment) that the source data has blocks with other block_types in addition to problem types.  Then, the final assertion should loop through all returned blocks to show that they are all problems.  As written, it feels like this would bass even if the block_types parameter was a no-op, unless I'm missing something.

nasthagiri replied:
Agree.  We want to test that non-problem blocks are NOT in the returned list, while problem blocks are.

If you move this test method to `test_serializers.py`, there you can test that it works as expected for both serializers by adding a common helper in `TestBlockSerializerBase` that is called by methods in `TestBlockSerializer` and `TestBlockDictSerializer`.  You can do so by looping through all the blocks in `self.block_structure` and making sure that only blocks of type `problem` are returned by the serializer.

robrap commented on openedx/core/lib/block_structure/block_structure.py:
```diff
@@ -66,11 +66,11 @@ def __init__(self, root_block_usage_key):
-        need to support DAGs.
+        The default iterator for a block structure is get_block_keys()
+        since we need to filter blocks as a list.
+        A topological traversal can be used to support DAGs.
```
@nasthagiri Can you help clarify when you get the chance? Is a block_structure a block_dag?  My understanding was that the results of this process will no longer be a block_structure, but the transformers still need it.  Are we introducing a new block_list or something like that?  I'm afraid I'm not being too helpful here without you.

Thanks again @lenacom.  Hopefully @nasthagiri can help clarify soon.

nasthagiri replied:
@robrap Yes, a block_structure remains as a block_dag.  However, this change here is to change the default iterator (which is used only by the Course API serializer) to iterate through all the block keys in the structure.  All Block Transformers, on the other hand, do not use this iterator.  They will continue to be explicit about their traversal requirements: topological, post_order, etc.

nasthagiri commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -61,6 +64,16 @@ def get_blocks(
+            if block_type not in block_types:
+                block_keys_to_remove.append(block_key)
+        for block_key in block_keys_to_remove:
+            blocks.remove_block(block_key, True)
```
For code readability, please add the keyword name to the boolean parameter: `keep_descendants=True`.

nasthagiri commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -85,6 +86,11 @@ class BlocksView(DeveloperErrorViewMixin, ListAPIView):
 
           Example: return_type=dict
 
+        * block_types: (list) Requested types of blocks. Possible values include sequential,
```
Can you expand this docstring to indicate that this parameter is actually used to "filter" the response?

nasthagiri replied:
What do we think about changing this name to `block_type_filter`?

nasthagiri commented:
@lenacom I have completed my pass.  It looks good overall.
Primary changes required:
* Resurrect `root` - it is needed for determining the root of a block structure.
* Rename `block_types` to `block_type_filter`.
* Move the `block_types` test method to `serializers.py` and make sure both positive (problem-type) and negative (non-problem-type) tests exist.

lenacom commented:
jenkins run js

lenacom commented:
jenkins run bokchoy

lenacom commented:
@nasthagiri I implemented the comments (restored 'root', renamed 'block_types' to 'block_type_filter', changed doc strings), I only didn't understand how to write a test on 'block_type_filter' in test_serializer.py because filtering is done before creating a serializer. I improved the test on filtering in test_api.py. Please review.

nasthagiri commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -85,6 +86,12 @@ class BlocksView(DeveloperErrorViewMixin, ListAPIView):
           Example: return_type=dict
 
+        * block_type_filter: (list) Requested types of blocks used to filter the final result
+          of returned blocks. Possible values include sequential,vertical, html, problem,
```
Nit: missing space before "vertical".

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_api.py:
```diff
@@ -77,3 +79,21 @@ def test_sub_structure(self):
+        blocks = get_blocks(self.request, sequential_block.location, self.user)
+        self.assertEquals(len(blocks['blocks']), 5)
+        found_not_problem = False
+        for key in blocks['blocks']:
```
Nit: Instead of using a regex, do an equality check against `block['type']`.
So:
```
for block in blocks['blocks']:
    if block['type'] != 'problem'
```

nasthagiri commented on lms/djangoapps/course_api/blocks/api.py:
```diff
@@ -61,6 +64,16 @@ def get_blocks(
+        block_keys_to_remove = []
+        for block_key in blocks:
+            block_type = blocks.get_xblock_field(block_key, 'category')
+            if block_type not in block_type_filter:
```
Hmm.. reading this line of code makes me 2nd guess the naming of the filter variable.  It's unclear from the name that it's a list.  Sorry for one more renaming, but how about changing it to: `block_types_filter` (with an "s")?

nasthagiri commented:
@lenacom This looks great.  Thanks.  I have just a few Nit comments.

lenacom commented:
jenkins run bokchoy

lenacom commented:
@nasthagiri I implemented the requested changes.

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_api.py:
```diff
@@ -77,3 +77,22 @@ def test_sub_structure(self):
+        blocks = get_blocks(self.request, sequential_block.location, self.user, requested_fields=['type'])
+        self.assertEquals(len(blocks['blocks']), 5)
+        found_not_problem = False
+        for key in blocks['blocks']:
```
Nit: Python provides a nice utility (`iteritems`) for iterating through all key, value pairs in a dict, to avoid the performance hit of retrieving the item again in your loop.  So it would be:
```
for key, block in blocks['blocks'].iteritems()
    if block['type'] != 'problem'
```
And, since you don't even need the `key` in the loop, you can just use `itervalues`:
```
for block in blocks['blocks'].itervalues()
    if block['type'] != 'problem'
```

nasthagiri commented:
@lenacom Once again, thanks for your contribution and your patience through this process.  The PR looks good from my end.
:+1:

lenacom commented:
jenkins run bokchoy

robrap commented:
@douglashall - Here's the PR.  
@lenacom - We are going to test this on our end.

douglashall commented:
Tested the LTI URLs returned by the API with a local installation of Canvas and things are working well. Nice work @lenacom!

Couple of questions for you and @robrap @nasthagiri?
Should course blocks exclude the lti_url field since the course is not a valid LTI provider integration point?
Should we update the [doc](http://edx.readthedocs.org/projects/open-edx-building-and-running-a-course/en/latest/course_features/lti/lti_address_content.html) related to finding LTI URLs to include this method?

nasthagiri commented:
Please be mindful that there's another OSPR that is changing the same Course Blocks API files: https://github.com/edx/edx-platform/pull/11664

So there will be a merge conflict depending on who merges first.

nasthagiri commented:
@douglashall I spoke with @robrap about this.

We decided to keep the PR as it is, for now.  Yes, it's true that some block types (e.g., courses) may not be supported via an LTI provider integration point.  However, it's also true that some courses may not have LTI integration enabled and so all blocks within that course would have invalid LTI URLs.  

So the LTI URLs provided by this API will not always work.  It will be up to the caller to handle the case when the URL raises an error.  We should address this in the documentation.

douglashall commented:
@nasthagiri Sounds good to me. :+1:

robrap commented:
@catong We'll need a story to update the docs regarding this feature and its limitations.  Let me know how I can help.  Thanks!

catong commented:
@robrap https://openedx.atlassian.net/browse/DOC-2793

robrap commented:
Thanks @catong
